File: src/solar/solar-common/util/system/MessageUtil.ts
import ActorItem from "@/ActorItem";
import ActorAbility from "@/ActorAbility";
import DataBase from "@/DataBase";

export default class MessageUtil {

    static hookEventCallBacks: {
        [type: string]: ((msg: any) => boolean)[]
    } = {}

    /** 返回true 表示忽略此次点击  eventId 1=左键点击*/
    static windowEventCallBacks: ((this: void, eventId: number) => boolean)[] = [];


    private static _sl_isInitialized = false

    // private static _sl_lastReleaseItem = null

    /**
     *
     * @param callBack 返回true 表示忽略此次点击
     */
    static addWindowEventCallBack(callBack: ((this: void, eventId: number) => boolean)) {
        MessageUtil.windowEventCallBacks.push(callBack)
    }


    private static _sl_onHookMsg(msg: any): boolean {
        let type = msg.type;
        if (type == null) {
            return true
        }
        if (type == "key_down") {
            // MessageUtil._sl_lastReleaseItem = null
        } else if (type == "mouse_ability") {//鼠标点击技能
            if (msg.ability == null) {
                return true
            }
            let actor: ActorAbility = DataBase.getAbilityTypeSolarData(id2string(msg.ability), false)?._SL_solarActorAbility;
            if (actor != null) {
                if (actor.isPassive() || actor.isDisable() || actor.isHide()) {
                    return false
                }
                //目前只能监听到左键点击技能
                let b = actor.localClick(msg.code == 4 ? 2 : 1, msg.x, msg.y);
                if (b == false) {
                    return false;
                }
            }
        } else if (type == "mouse_item") {
            //点击技能时 msg.ability 为0
            let order: number = msg.order;
            let itemIndex = order - 852008
            // if (MessageUtil._sl_lastReleaseItem != null) {
            //     MessageUtil._sl_lastReleaseItem = null
            //     return true
            // }
            // if (msg.code == 4) {
            //     MessageUtil._sl_lastReleaseItem = itemIndex
            //     return true;
            // }
            let item = UnitItemInSlot(selection(), itemIndex);
            let actor: ActorItem = DataBase.getItemSolarData(item, false)?._SL_solarActorItem;
            if (actor != null) {
                //左键点击被动物品 拦截 (模拟被动物品 在准备施放技能时发布停止命令即可 这里拦截可能会导致 选择物品目标 交换物品位置不可用)
                // if (msg.code == 1 && actor.isPassive() || actor.isDisable() || actor.isHide()) {
                //     return false
                // }
                let b = actor.localClick(msg.code == 4 ? 2 : 1, msg.x, msg.y);
                //返回false 可能会导致 选择物品目标 交换物品位置不可用
                if (b == false) {
                    return false;
                }
            }
        }

        let hookEventCallBacks = MessageUtil.hookEventCallBacks[type];
        if (hookEventCallBacks == null) {
            return true;
        }
        for (let callBack of hookEventCallBacks) {
            let f = callBack(msg);
            if (f == false) {
                return f
            }
        }
        return true;
    }

    static _sl_init_message_hook() {
        if (MessageUtil._sl_isInitialized) {
            return;
        }
        MessageUtil._sl_isInitialized = true;
        //
        if (isEmbedJapi == false && isDebug) {
            print("提示：部分we注册message.hook事件会导致释放技能崩溃!(标准ydwe不会！)")
        }
        //
        const message: NoSelf = require('jass.message')
        let hook: (this: void, msg) => boolean = message.hook;
        if (hook == null) {
            message.hook = function (msg: any) {
                return MessageUtil._sl_onHookMsg(msg);
            }
        } else {
            message.hook = function (this: void, msg: any) {
                let flag = MessageUtil._sl_onHookMsg(msg);
                if (flag == false) {
                    //返回fasle 表示拦截
                    return flag;
                }
                return hook(msg);
            }
        }
        //窗口根事件 可以在这里做一些比message.hook 还要提前的事件处理
        if (isEmbedJapi) {
            let oldWindowEventCallBack: (this: void, eventId: number) => boolean = _G["WindowEventCallBack"];
            if (oldWindowEventCallBack == null) {
                _G["WindowEventCallBack"] = function (this: void, eventId: number) {
                    for (let windowEventCallBack of MessageUtil.windowEventCallBacks) {
                        let flag = windowEventCallBack(eventId);
                        if (flag == true) {
                            return true;
                        }
                    }
                    return false;
                }
            } else {
                _G["WindowEventCallBack"] = function (this: void, eventId: number) {
                    for (let windowEventCallBack of MessageUtil.windowEventCallBacks) {
                        let flag = windowEventCallBack(eventId);
                        if (flag == true) {
                            return true;
                        }
                    }
                    return oldWindowEventCallBack(eventId);
                }
            }
        }

    }

}

File: scripts/solar/solar-common/util/system/MessageUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["6"] = 3,["7"] = 3,["8"] = 5,["9"] = 5,["10"] = 5,["12"] = 5,["13"] = 23,["14"] = 24,["15"] = 24,["16"] = 23,["17"] = 28,["18"] = 29,["19"] = 30,["20"] = 31,["22"] = 33,["23"] = 35,["24"] = 36,["25"] = 37,["27"] = 39,["28"] = 39,["29"] = 39,["30"] = 39,["31"] = 39,["32"] = 40,["33"] = 41,["34"] = 42,["36"] = 45,["37"] = 46,["38"] = 47,["41"] = 50,["42"] = 52,["43"] = 53,["44"] = 62,["45"] = 62,["46"] = 62,["47"] = 62,["48"] = 63,["49"] = 63,["50"] = 64,["51"] = 69,["52"] = 71,["53"] = 72,["57"] = 77,["58"] = 78,["59"] = 79,["61"] = 81,["62"] = 82,["63"] = 83,["64"] = 84,["67"] = 87,["68"] = 28,["69"] = 90,["70"] = 91,["73"] = 94,["74"] = 96,["75"] = 97,["77"] = 100,["78"] = 101,["79"] = 102,["80"] = 103,["81"] = 104,["82"] = 103,["84"] = 107,["85"] = 108,["86"] = 109,["87"] = 111,["89"] = 113,["90"] = 107,["92"] = 117,["93"] = 118,["94"] = 119,["95"] = 120,["96"] = 121,["97"] = 122,["98"] = 123,["99"] = 124,["102"] = 127,["103"] = 120,["105"] = 130,["106"] = 131,["107"] = 132,["108"] = 133,["109"] = 134,["112"] = 137,["113"] = 130,["116"] = 90,["117"] = 9,["118"] = 12,["119"] = 15});
local ____exports = {}
local ____DataBase = require("solar.solar-common.common.DataBase")
local DataBase = ____DataBase.default
____exports.default = __TS__Class()
local MessageUtil = ____exports.default
MessageUtil.name = "MessageUtil"
function MessageUtil.prototype.____constructor(self)
end
function MessageUtil.addWindowEventCallBack(self, callBack)
    local ____exports_default_windowEventCallBacks_0 = ____exports.default.windowEventCallBacks
    ____exports_default_windowEventCallBacks_0[#____exports_default_windowEventCallBacks_0 + 1] = callBack
end
function MessageUtil._sl_onHookMsg(self, msg)
    local ____type = msg.type
    if ____type == nil then
        return true
    end
    if ____type == "key_down" then
    elseif ____type == "mouse_ability" then
        if msg.ability == nil then
            return true
        end
        local ____opt_1 = DataBase:getAbilityTypeSolarData(
            id2string(msg.ability),
            false
        )
        local actor = ____opt_1 and ____opt_1._SL_solarActorAbility
        if actor ~= nil then
            if actor:isPassive() or actor:isDisable() or actor:isHide() then
                return false
            end
            local b = actor:localClick(msg.code == 4 and 2 or 1, msg.x, msg.y)
            if b == false then
                return false
            end
        end
    elseif ____type == "mouse_item" then
        local order = msg.order
        local itemIndex = order - 852008
        local item = UnitItemInSlot(
            selection(),
            itemIndex
        )
        local ____opt_3 = DataBase:getItemSolarData(item, false)
        local actor = ____opt_3 and ____opt_3._SL_solarActorItem
        if actor ~= nil then
            local b = actor:localClick(msg.code == 4 and 2 or 1, msg.x, msg.y)
            if b == false then
                return false
            end
        end
    end
    local hookEventCallBacks = ____exports.default.hookEventCallBacks[____type]
    if hookEventCallBacks == nil then
        return true
    end
    for ____, callBack in ipairs(hookEventCallBacks) do
        local f = callBack(nil, msg)
        if f == false then
            return f
        end
    end
    return true
end
function MessageUtil._sl_init_message_hook(self)
    if ____exports.default._sl_isInitialized then
        return
    end
    ____exports.default._sl_isInitialized = true
    if isEmbedJapi == false and isDebug then
        print("提示：部分we注册message.hook事件会导致释放技能崩溃!(标准ydwe不会！)")
    end
    local message = require("jass.message")
    local hook = message.hook
    if hook == nil then
        message.hook = function(msg)
            return ____exports.default:_sl_onHookMsg(msg)
        end
    else
        message.hook = function(msg)
            local flag = ____exports.default:_sl_onHookMsg(msg)
            if flag == false then
                return flag
            end
            return hook(msg)
        end
    end
    if isEmbedJapi then
        local oldWindowEventCallBack = _G.WindowEventCallBack
        if oldWindowEventCallBack == nil then
            _G.WindowEventCallBack = function(eventId)
                for ____, windowEventCallBack in ipairs(____exports.default.windowEventCallBacks) do
                    local flag = windowEventCallBack(eventId)
                    if flag == true then
                        return true
                    end
                end
                return false
            end
        else
            _G.WindowEventCallBack = function(eventId)
                for ____, windowEventCallBack in ipairs(____exports.default.windowEventCallBacks) do
                    local flag = windowEventCallBack(eventId)
                    if flag == true then
                        return true
                    end
                end
                return oldWindowEventCallBack(eventId)
            end
        end
    end
end
MessageUtil.hookEventCallBacks = {}
MessageUtil.windowEventCallBacks = {}
MessageUtil._sl_isInitialized = false
return ____exports
